src/provider/mlkzg.rs

```diff
@@ -321,21 +311,39 @@ where
       // let R0 = -W[0];
       // let R1 = -W[1];
       // let R2 = -W[2];
-      // let L = L0 + L1*d[0] + L2*d[1];
-      // let R = R0 + R1*d[0] + R2*d[1];
+      // let L = L0 + L1*d_0 + L2*d_1;
+      // let R = R0 + R1*d_0 + R2*d_1;
+      //
+      // Note, that intermediate computation of C_B can be replaced by MSM of C with the
+      // powers multiplied by (1 + d_0 + d_1)
       //
       // We group terms to reduce the number of scalar mults (to seven):
       // In Rust, we could use MSMs for these, and speed up verification.
-      let L = E::G1::from(C_B) * (E::Fr::ONE + d[0] + d[1])
-        - E::G1::from(vk.g) * (B_u[0] + d[0] * B_u[1] + d[1] * B_u[2])
+
+      let q_power_multiplier = E::Fr::ONE + d_0 + d_1;
+
+      let q_powers_multiplied: Vec<E::Fr> = q_powers
+        .iter()
+        .map(|q_power| *q_power * q_power_multiplier)
+        .collect();
+
+      // Compute the batched openings
+      // compute B(u_i) = v[i][0] + q*v[i][1] + ... + q^(t-1) * v[i][t-1]
+      let B_u = v
+        .iter()
+        .map(|v_i| zip_with!(iter, (q_powers, v_i), |a, b| *a * *b).sum())
+        .collect::<Vec<E::Fr>>();
+
+      let L = NE::GE::vartime_multiscalar_mul(&q_powers_multiplied[..k], &C[..k])
+        - E::G1::from(vk.g) * (B_u[0] + d_0 * B_u[1] + d_1 * B_u[2])
         + E::G1::from(W[0]) * u[0]
-        + E::G1::from(W[1]) * (u[1] * d[0])
-        + E::G1::from(W[2]) * (u[2] * d[1]);
+        + E::G1::from(W[1]) * (u[1] * d_0)
+        + E::G1::from(W[2]) * (u[2] * d_1);
```

Comment:
`L` can be computed using a single call to `vartime_multiscalar_mul`, with inputs
- `[q_powers_multiplied[..k], vk.g, W[0], W[1], W[2]]`
- `[C[..k], (B_u[0] + d_0 * B_u[1] + d_1 * B_u[2]), u[0], (u[1] * d_0), (u[2] * d_1)]`

In the context where we are writing the verifier can use an MSM implementation like in Solidity, this should be a bit more efficient than computing individual scalar multiplications .

Reply:
Yes, indeed! Added this refactoring in https://github.com/lurk-lab/arecibo/pull/231/commits/e1a5b159371e52858cd9f76c9278b9db32722933